lib/discourse_theme/cli.rb

```diff
@@ -167,6 +183,94 @@ def run(args)
         else
           UI.info "Manage: #{client.root}/admin/customize/themes/#{theme_id}"
         end
+      elsif command == "rspec"
+        index = dir.index("/spec")
+        spec_path = "/spec"
+
+        if index
+          spec_path = dir[index..-1]
+          dir = dir[0..index - 1]
+        end
+
+        raise DiscourseTheme::ThemeError.new "'#{dir} does not exist" unless Dir.exist?(dir)
+
+        # Checks if the container is running
+        container_name = DISCOURSE_TEST_DOCKER_CONTAINER_NAME
+        is_running = false
+        container_exists = false
+
+        if !(
+             output =
+               execute(command: "docker ps -a --format '{{json .}}' | grep #{container_name}")
+           ).empty?
+          container_exists = true
+          is_running = JSON.parse(output)["State"] == "running"
+        end
+
+        basename = Pathname.new(dir).basename.to_s
+        verbose = !!args.delete("--verbose")
+        headless = !!args.delete("--headless")
+
+        if !is_running || args.delete("--rebuild")
+          if container_exists
+            execute(command: "docker stop #{container_name}")
+            execute(command: "docker rm -f #{container_name}")
+          end
+
+          execute(
+            command: <<~CMD.squeeze(" "),
+              docker run -d \
+                -p 31337:31337 \
+                --add-host host.docker.internal:host-gateway \
+                --entrypoint=/sbin/boot \
+                --name=#{container_name} \
+                -v #{dir}:/tmp/#{basename} \
+                discourse/discourse_test:release
```

Comment:
`docker run {blah}` does not guarantee to use the latest version of the image. If the workstation already has an old copy, that will be used instead.

So I think the best strategy is probably to do what our `./launcher` script does in production: run `docker pull discourse/discourse_test:release` before the `docker run`

Reply:
This is a good point, I do feel like running `docker pull` is going to add significant overhead so I think I'll go with a versioned approach. In other words, for this version of the gem we will run this version of the `discourse/discourse_test:release` image.

Reply:
Btw I intend to follow up with this in another PR while I figure out if we have versioning for our images.

Reply:
We don't version our discourse_test image (tags are [here](https://hub.docker.com/r/discourse/discourse_test/tags))

The good news is that `docker pull` is basically free if you already have the latest image. It compares the image hashes of the local/remote versions before doing anything.

```
❯ time docker pull discourse/discourse_test:release
release: Pulling from discourse/discourse_test
Digest: sha256:2a78135423931fe11fd7edfccf19c476125a1fbc0e12cabf56cf79c2405877f8
Status: Image is up to date for discourse/discourse_test:release
docker.io/discourse/discourse_test:release
docker pull discourse/discourse_test:release

0.08s user
0.06s system
10% cpu
1.409 total
```

I don't think we should 'release' this without the `docker pull`. If we do, we're gonna have a world of pain when people start reporting weird issues caused by out-of-date images.

Reply:
I've added `--pull=always` when starting the container to resolve this :+1: